### PR TITLE
[Rating] Add titles for each stars

### PR DESCRIPTION
**Pull request by lepit31**
Add an attribute "titles"  to set a custom  title for each star by it position on array.
if the index of  the star is greater than the titles array length return the  index +1 of the star
Add a default value ["bad","poor","regular","good","gorgeous"]

### CODE REVIEW

**Comment by karianna** [comment]
@lepit31 - The existing tests fail and we'd also like to see some new tests for this please :-).

**Comment by lepit31** [comment]
Fix the compilation bug 
Add 5 test for the rating
(5 test are ok  with a 'grunt test' on my local repository )

This is my first  git commit , sorry for the annoyance.

**Comment by rvanbaalen** [comment]
No problem man. Appreciate the input. Everybody has got to start somewhere.

File: src/rating/docs/readme.md [hunk 1]
```diff
@@ -16,6 +16,10 @@ Rating directive that will take care of visualising a star rating bar.
  	_(Defaults: false)_ :
  	Prevent user's interaction.
 
+ * `titles`
+ 	_(Defaults: ["bad","poor","regular","good","gorgeous"])_ :
```
Comment (wesleycho) [inline src/rating/docs/readme.md]: Add spaces after each comma

File: src/rating/rating.js [hunk 2]
```diff
@@ -3,7 +3,8 @@ angular.module('ui.bootstrap.rating', [])
 .constant('ratingConfig', {
   max: 5,
   stateOn: null,
-  stateOff: null
+  stateOff: null,
+  titles : ['bad','poor','regular','good','gorgeous']
```
Comment (wesleycho) [inline src/rating/rating.js]: Add spaces after each comma

File: src/rating/rating.js [hunk 3]
```diff
@@ -22,7 +23,9 @@ angular.module('ui.bootstrap.rating', [])
 
     this.stateOn = angular.isDefined($attrs.stateOn) ? $scope.$parent.$eval($attrs.stateOn) : ratingConfig.stateOn;
     this.stateOff = angular.isDefined($attrs.stateOff) ? $scope.$parent.$eval($attrs.stateOff) : ratingConfig.stateOff;
-
+    var tmpTitles = angular.isDefined($attrs.titles)  ? $scope.$parent.$eval($attrs.titles) : ratingConfig.titles ;    
+    this.titles =   angular.isArray(tmpTitles) && tmpTitles.length > 0 ? tmpTitles  : ratingConfig.titles;
```
Comment (wesleycho) [inline src/rating/rating.js]: Delete extra whitespace before `angular.isArray` and delete one space before `:`

File: src/rating/rating.js [hunk 4]
```diff
@@ -30,11 +33,20 @@ angular.module('ui.bootstrap.rating', [])
 
   this.buildTemplateObjects = function(states) {
     for (var i = 0, n = states.length; i < n; i++) {
-      states[i] = angular.extend({ index: i }, { stateOn: this.stateOn, stateOff: this.stateOff }, states[i]);
+      states[i] = angular.extend({ index: i }, { stateOn: this.stateOn, stateOff: this.stateOff, title:this.getTitle(i) }, states[i]);
```
Comment (wesleycho) [inline src/rating/rating.js]: Add space after `title:`

File: src/rating/rating.js [hunk 5]
```diff
@@ -30,11 +33,20 @@ angular.module('ui.bootstrap.rating', [])
...
     }
     return states;
   };
-
+  
+  this.getTitle = function (index) {
```
Comment (wesleycho) [inline src/rating/rating.js]: Delete space before `(`

File: src/rating/rating.js [hunk 6]
```diff
@@ -30,11 +33,20 @@ angular.module('ui.bootstrap.rating', [])
...
+	  if( index >= this.titles.length) {
```
Comment (wesleycho) [inline src/rating/rating.js]: Fix spacing of if statement, i.e.
```
if (index >= this.titles.length) {
```

File: src/rating/rating.js [hunk 7]
```diff
@@ -30,11 +33,20 @@ angular.module('ui.bootstrap.rating', [])
...
+		  return index+1;
```
Comment (wesleycho) [inline src/rating/rating.js]: Change to `return index + 1`

File: src/rating/test/rating.spec.js [hunk 8]
```diff
@@ -27,6 +27,15 @@ describe('rating directive', function () {
     return state;
   }
 
+    function getTitles() {
```
Comment (wesleycho) [inline src/rating/test/rating.spec.js]: Fix indentation (if it is a tab character, change to spaces)

File: src/rating/test/rating.spec.js [hunk 9]
```diff
@@ -27,6 +27,15 @@ describe('rating directive', function () {
...
+    var stars = getStars();
+    var titles = [];
+    for (var i = 0, n = stars.length; i < n; i++) {
+      titles.push( stars.eq(i).attr('title') );
+    }
```
Comment (wesleycho) [inline src/rating/test/rating.spec.js]: Change for loop to
```
var titles = stars.map(function(idx, star) {
  return star.attr('title');
});
```

File: src/rating/test/rating.spec.js [hunk 10]
```diff
@@ -267,4 +276,48 @@ describe('rating directive', function () {
       expect(getState('on', 'off')).toEqual([true, true, true, true, true, false, false, false, false, false]);
     });
   });
+  
+  describe('Default title', function() {
+    it('should return the default title for each star', function() {
+      expect(getTitles()).toEqual(['bad','poor','regular','good','gorgeous']);
```
Comment (wesleycho) [inline src/rating/test/rating.spec.js]: Fix spacing of array elements

File: src/rating/test/rating.spec.js [hunk 11]
```diff
@@ -267,4 +276,48 @@ describe('rating directive', function () {
...
+      element = $compile('<rating ng-model="rate"></rating>')($rootScope);
+      $rootScope.$digest();
+    }));
+    afterEach(inject(function(ratingConfig) {
+      // return it to the original state
+      angular.extend(ratingConfig, originalConfig);
+    }));
+ 
+	 it('should return the default title for each star', function() {
+		  expect(getTitles()).toEqual(['bad','poor','regular','good','gorgeous','6','7','8','9','10']);
```
Comment (wesleycho) [inline src/rating/test/rating.spec.js]: Fix spacing of array elements

File: src/rating/test/rating.spec.js [hunk 12]
```diff
@@ -267,4 +276,48 @@ describe('rating directive', function () {
...
+		});
+  });
+  
+    describe('shows custom titles ', function() {
```
Comment (wesleycho) [inline src/rating/test/rating.spec.js]: Fix indentation

File: src/rating/test/rating.spec.js [hunk 13]
```diff
@@ -267,4 +276,48 @@ describe('rating directive', function () {
...
+		it('should return the custom title for each star', function() {
+			element = $compile('<rating ng-model="rate" titles="[44,45,46]"></rating>')($rootScope);
+			$rootScope.$digest();
+		  expect(getTitles()).toEqual(['44','45','46','4','5']);
```
Comment (wesleycho) [inline src/rating/test/rating.spec.js]: Fix spacing

File: src/rating/test/rating.spec.js [hunk 14]
```diff
@@ -267,4 +276,48 @@ describe('rating directive', function () {
...
+		});
+		it('should return the default title if the custom title is an empty array', function() {
+			element = $compile('<rating ng-model="rate" titles="[]"></rating>')($rootScope);
+			$rootScope.$digest();
+		  expect(getTitles()).toEqual(['bad','poor','regular','good','gorgeous']);
```
Comment (wesleycho) [inline src/rating/test/rating.spec.js]: Fix spacing

File: src/rating/test/rating.spec.js [hunk 15]
```diff
@@ -267,4 +276,48 @@ describe('rating directive', function () {
...
+  });
+  
+    describe('shows custom titles ', function() {
+		it('should return the custom title for each star', function() {
+			element = $compile('<rating ng-model="rate" titles="[44,45,46]"></rating>')($rootScope);
+			$rootScope.$digest();
+		  expect(getTitles()).toEqual(['44','45','46','4','5']);
+		});
+		it('should return the default title if the custom title is an empty array', function() {
+			element = $compile('<rating ng-model="rate" titles="[]"></rating>')($rootScope);
```
Comment (wesleycho) [inline src/rating/test/rating.spec.js]: This test should omit the titles attribute completely

File: src/rating/test/rating.spec.js [hunk 16]
```diff
@@ -267,4 +276,48 @@ describe('rating directive', function () {
...
+			$rootScope.$digest();
+		  expect(getTitles()).toEqual(['bad','poor','regular','good','gorgeous']);
+		});
+	 it('should return the default title if the custom title is not an array', function() {
```
Comment (wesleycho) [inline src/rating/test/rating.spec.js]: Fix indentation

**Comment by wesleycho** [comment]
I spent some time refactoring the work & cleaning up the history just now - I changed the default values to `['one', 'two', 'three', 'four', 'five']`, since it is more suitably generic. I will be merging shortly.

**Pull request closed without merging**